bindings/pyroot_experimental/PyROOT/python/ROOT/pythonization/_rtensor.py

```diff
@@ -0,0 +1,110 @@
+# Author: Stefan Wunsch CERN  07/2019
+
+################################################################################
+# Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.                      #
+# All rights reserved.                                                         #
+#                                                                              #
+# For the licensing terms see $ROOTSYS/LICENSE.                                #
+# For the list of contributors see $ROOTSYS/README/CREDITS.                    #
+################################################################################
+
+from ROOT import pythonization
+from libROOTPython import GetEndianess, GetTensorDataPointer, GetSizeOfType, AsRTensor
+from ROOT.pythonization._rvec import _array_interface_dtype_map
+import cppyy
+
+
+def get_array_interface(self):
+    cppname = type(self).__cppname__
+    for dtype in _array_interface_dtype_map:
+        if not cppname.find("RTensor<{},".format(dtype)) is -1:
+            dtype_numpy = _array_interface_dtype_map[dtype]
+            dtype_size = GetSizeOfType(dtype)
+            endianess = GetEndianess()
+            shape = self.GetShape()
+            strides = self.GetStrides()
+            # Numpy breaks for data pointer of 0 even though the array is empty.
+            # We set the pointer to 1 but the value itself is arbitrary and never accessed.
+            pointer = GetTensorDataPointer(self, cppname)
+            if pointer == 0:
+                pointer == 1
+            return {
+                "shape": tuple(s for s in shape),
+                "strides": tuple(s * dtype_size for s in strides),
+                "typestr": "{}{}{}".format(endianess, dtype_numpy, dtype_size),
+                "version": 3,
+                "data": (pointer, False)
+            }
+
+
+def add_array_interface_property(klass, name):
+    if True in [
+            not name.find("RTensor<{},".format(dtype)) is -1 for dtype in _array_interface_dtype_map
+    ]:
+        klass.__array_interface__ = property(get_array_interface)
+
+
+def RTensorGetitem(self, idx):
+    # Make single index iterable and convert to list
+    if not hasattr(idx, "__len__"):
+        idx = [idx]
+    idx = list(idx)
+
+    # Check shape
+    shape = self.GetShape()
+    if shape.size() != len(idx):
+        raise Exception("RTensor with rank {} got {} indices.".format(shape.size(), len(idx)))
+
+    # Convert negative indices and Nones
+    isSlice = False
+    for i, x in enumerate(idx):
+        if type(x) == slice:
+            isSlice = True
```

Comment:
I am not sure if it would be useful here, but when I implemented slicing for `TSeqCollection` I used this:

https://github.com/root-project/root/blob/master/bindings/pyroot_experimental/PyROOT/python/ROOT/pythonization/_tseqcollection.py#L63

You can ask for the indices that the slice syntax represents (returns a start,stop,step tuple). Could this simplify the logic you have here and potentially allow step size > 1?

Reply:
The step size > 1 is not supported by the C++ implementation in the back and therefore it's excluded. Then, I have to convert the slice to the C++ representation of the slice (which is a `vector<vector<size_t>>`) taking into account the correct ranges.